# Classification Evaluation Metrics

Evaluating classification models requires metrics that go beyond simple accuracy, especially when dealing with imbalanced datasets. This notebook covers:
*   **Confusion Matrix:** Understanding True Positives, True Negatives, False Positives, and False Negatives.
*   **Precision & Recall:** Trade-offs between target class coverage and prediction reliability.
*   **F1-Score:** Harmonic mean of precision and recall.
*   **Accuracy:** Overall prediction correctness rate.

In [3]:
#Importing Libraries
# from sklearn.metrics import confusion_matrix
# from sklearn.metrics import precision_score
# from sklearn.metrics import recall_score
# from sklearn.metrics import accuracy_score
# from sklearn.metrics import f1_score

from sklearn.metrics import confusion_matrix, precision_score, recall_score, accuracy_score, f1_score

In [4]:
# Creating sample data

# Actual Output
y_true = [1,0,1,1,0,1,0,0,1,0,1,0,1,0,1]

# Predicted Output
y_pred = [1,0,1,0,0,1,1,0,1,0,0,0,1,0,1]

In [6]:
#Confusion matrix
cm = confusion_matrix(y_true, y_pred)

print("Confusion Matrix")
print(cm)

TN, FP, FN, TP = cm.ravel()

print("True Negative :", TN)
print("False Positive:", FP)
print("False Negative:", FN)
print("True Positive :", TP)

Confusion Matrix
[[6 1]
 [2 6]]
True Negative : 6
False Positive: 1
False Negative: 2
True Positive : 6


Conclusion:
 Out of all the samples, 80% of the samples are correctly predicted and the remaining 20% of the samples are incorrectly predicted. Hence, the model performance is good. 

The model correctly predicted 12 samples and wrongly predicted 3 samples.

Hence, the model performance is good.

In [7]:
#Precision Score (Positive Label)
precision_positive = precision_score(y_true, y_pred, pos_label=1)

print("Precision Score (Positive) =", precision_positive)
print("Percentage =", precision_positive*100,"%")

Precision Score (Positive) = 0.8571428571428571
Percentage = 85.71428571428571 %


Conclusion:
Out of all the predicted Positive labels,

85.71% samples are correctly predicted as Positive.

Remaining 14.29% are False Positives.

In [13]:
#Precision Score (Negative Label)
precision_negative = precision_score(y_true, y_pred, pos_label=0)

print("Precision Score (Negative) =", precision_negative)
print("Percentage =", precision_negative*100,"%")

Precision Score (Negative) = 0.75
Percentage = 75.0 %


Conclusion:
Out of all the predicted Negative labels,

75% samples are correctly predicted as Negative.

Remaining 25% are False Negatives.

In [8]:
#Recall Score (Positive Label)
recall_positive = recall_score(y_true, y_pred, pos_label=1)

print("Recall Score (Positive) =", recall_positive)
print("Percentage =", recall_positive*100,"%")

Recall Score (Positive) = 0.75
Percentage = 75.0 %


Conclusion
Out of all the actual Positive labels,

75% samples are correctly predicted as Positive.

Remaining 25% are False Negatives.

In [9]:
#Recall Score (Negative Label)
recall_negative = recall_score(y_true, y_pred, pos_label=0)

print("Recall Score (Negative) =", recall_negative)
print("Percentage =", recall_negative*100,"%")

Recall Score (Negative) = 0.8571428571428571
Percentage = 85.71428571428571 %


Conclusion
Out of all the actual Negative labels,

85.71% samples are correctly predicted as Negative.

Remaining 14.29% are False Positives.

In [10]:
#Accuracy Score
accuracy = accuracy_score(y_true, y_pred)

print("Accuracy Score =", accuracy)
print("Percentage =", accuracy*100,"%")

Accuracy Score = 0.8
Percentage = 80.0 %


Conclusion:

The model correctly predicted 80% of the total samples.

Hence, the overall performance of the model is good.

In [11]:
#F1 Score (Positive Label)
f1_positive = f1_score(y_true, y_pred, pos_label=1)

print("F1 Score (Positive) =", f1_positive)
print("Percentage =", f1_positive*100,"%")

F1 Score (Positive) = 0.8
Percentage = 80.0 %


Conclusion:
The F1 Score for the Positive label is 80%.

It shows a good balance between Precision and Recall for the Positive class.

In [12]:
#F1 Score (Negative Label)
f1_negative = f1_score(y_true, y_pred, pos_label=0)

print("F1 Score (Negative) =", f1_negative)
print("Percentage =", f1_negative*100,"%")

F1 Score (Negative) = 0.8
Percentage = 80.0 %


Conclusion:
The F1 Score for the Negative label is 80%.

It shows a good balance between Precision and Recall for the Negative class.